clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,341 @@
+    private static final Logger log = LoggerFactory.getLogger(KafkaMetricsCollector.class);
+
+    private final StateLedger ledger;
+    private final Clock clock;
+    private final MetricNamingStrategy<MetricName> metricNamingStrategy;
+
+    private static final Field METRIC_VALUE_PROVIDER_FIELD;
+
+    static {
+        try {
+            METRIC_VALUE_PROVIDER_FIELD = KafkaMetric.class.getDeclaredField("metricValueProvider");
+            METRIC_VALUE_PROVIDER_FIELD.setAccessible(true);
+        } catch (Exception e) {
+            throw new KafkaException(e);
+        }
+    }
+
+    public KafkaMetricsCollector(MetricNamingStrategy<MetricName> metricNamingStrategy) {
+        this(metricNamingStrategy, Clock.systemUTC());
+    }
+
+    // Visible for testing
+    public KafkaMetricsCollector(MetricNamingStrategy<MetricName> metricNamingStrategy, Clock clock) {
+        this.metricNamingStrategy = metricNamingStrategy;
+        this.clock = clock;
+        this.ledger = new StateLedger();
+    }
+
+    public void init(List<KafkaMetric> metrics) {
+        ledger.init(metrics);
+    }
+
+    /**
+     * This is called whenever a metric is updated or added
+     */
+    public void metricChange(KafkaMetric metric) {
+        ledger.metricChange(metric);
+    }
+
+    /**
+     * This is called whenever a metric is removed
+     */
+    public void metricRemoval(KafkaMetric metric) {
+        ledger.metricRemoval(metric);
+    }
+
+    // Visible for testing
+    Set<MetricKey> getTrackedMetrics() {
+        return ledger.metricMap.keySet();
+    }
+
+    @Override
+    public void collect(MetricsEmitter metricsEmitter) {
+        for (Map.Entry<MetricKey, KafkaMetric> entry : ledger.getMetrics()) {
+            MetricKey metricKey = entry.getKey();
+            KafkaMetric metric = entry.getValue();
+
+            try {
+                collectMetric(metricsEmitter, metricKey, metric);
+            } catch (Exception e) {
+                // catch and log to continue processing remaining metrics
+                log.error("Unexpected error processing Kafka metric {}", metricKey, e);
+            }
+        }
+    }
+
+    protected void collectMetric(MetricsEmitter metricsEmitter, MetricKey metricKey, KafkaMetric metric) {
+        Object metricValue;
+
+        try {
+            metricValue = metric.metricValue();
+        } catch (Exception e) {
+            // If an exception occurs when retrieving value, log warning and continue to process the rest of metrics
+            log.warn("Failed to retrieve metric value {}", metricKey.name(), e);
+            return;
+        }
+
+        if (isMeasurable(metric)) {
+            Measurable measurable = metric.measurable();
+            Double value = (Double) metricValue;
+
+            if (measurable instanceof WindowedCount || measurable instanceof CumulativeSum) {
+                collectSum(metricKey, value, metricsEmitter);
+                collectDelta(metricKey, value, metricsEmitter);
```

Comment:
collectSum / collectDelta may produce additional metric payloads, which are not defined in the kip. We might want to add some tests to ensure that for the metrics we collect in the KIP we only get the payloads as we documented them.

Reply:
Currently this reports two different metrics with different names (e.g. one with `delta`, the other without). The KIP indicates we should only report one sum value per metric, which is either cumulative or delta based on the aggregation temporality requested by the broker, so this would have to respect that.

Reply:
I have created a single metric with either cumulative or delta temporality as per the subscription information received from broker (MetricsEmitter implementation should PR should cover that).